### PR TITLE
Improve allof

### PR DESCRIPTION
**Pull request by jianyexi**

### CODE REVIEW

File: src/lib/util/logging.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ function getTimeStamp() {
       return "0" + number
     }
 
-    return number
+    return number + ""
```
Comment (ruowan) [inline src/lib/util/logging.ts:22]: Why `+ ""`

Comment (jianyexi) [reply src/lib/util/logging.ts:22]: always return a string

File: src/lib/util/resolveSwagger.ts [hunk 2]
```diff
@@ -57,6 +57,25 @@ export function mergeArrays<T extends Data>(source: ReadonlyArray<T>, target: T[
   return target
 }
 
+function getParamKey(source: any) {
+  return source.$ref ? source.$ref : source.in + source.name
+}
+
+function getNextKey(source: sm.MutableStringMap<Data>) {
```
Comment (ruowan) [inline src/lib/util/resolveSwagger.ts:64]: add comment.

File: src/lib/util/resolveSwagger.ts [hunk 3]
```diff
@@ -93,6 +114,182 @@ export class ResolveSwagger {
...
+    }
+    const swagger = this.innerSwagger as any
+    const paths = swagger.paths
+    if (paths && paths instanceof Object && toArray(sm.keys(paths)).length > 0) {
+      for (const [property, v] of sm.entries(paths)) {
+        const pathsLevelParameters = (v as any).parameters
+        if (!pathsLevelParameters) {
+          continue
+        }
+        for (const [httpMethod, o] of sm.entries(v as any)) {
```
Comment (ruowan) [inline src/lib/util/resolveSwagger.ts:129]: add comment. renaming

File: src/lib/util/resolveSwagger.ts [hunk 4]
```diff
@@ -93,6 +114,182 @@ export class ResolveSwagger {
...
+          const ref = (s as any).$ref
+          referenceChain.push(ref)
+          const referredSchema = this.dereference(ref)
+          this.checkCircularAllOf(referredSchema, visited, referenceChain)
+          referenceChain.pop()
+        })
+    }
+  }
+
+  private getModelName(ref: string) {
```
Comment (ruowan) [inline src/lib/util/resolveSwagger.ts:253]: comments

File: src/lib/util/resolveSwagger.ts [hunk 5]
```diff
@@ -93,6 +114,182 @@ export class ResolveSwagger {
...
+    return undefined
+  }
+
+  private dereferenceInner(ref: string, visitedRefs: Set<string>): any {
+    const model = this.getModelName(ref)
+    if (model) {
+      if (visitedRefs.has(ref)) {
+        throw new Error("Circular reference")
+      }
+      if (visitedRefs.size > 40) {
```
Comment (ruowan) [inline src/lib/util/resolveSwagger.ts:267]: 40

File: src/lib/util/resolveSwagger.ts [hunk 6]
```diff
@@ -93,6 +114,182 @@ export class ResolveSwagger {
...
+      if (allOfSchema.$ref) {
+        allOfSchema = this.dereference(allOfSchema.$ref)
+        if (!allOfSchema) {
+          throw new Error("Invalid reference:" + allOfSchema.$ref)
+        }
+      }
+      if (allOfSchema.properties) {
+        sm.keys(allOfSchema.properties).forEach(key => {
+          if (sm.keys(schemaList).some(k => k === key)) {
+            if (!this.isEqual(schemaList[key], allOfSchema.properties[key])) {
```
Comment (ruowan) [inline src/lib/util/resolveSwagger.ts:178]: parameter order

File: src/lib/util/resolveSwagger.ts [hunk 7]
```diff
@@ -93,6 +114,182 @@ export class ResolveSwagger {
...
+          }
+        })
+        schema.required = requiredProperties
+      }
+    }
+    schema.properties = schemaList
+  }
+  /**
+   * @description Compare two property to see if they have same definitions.
+   * @param parentProperty
```
Comment (ruowan) [inline src/lib/util/resolveSwagger.ts:200]: comment what paramenterProperty mean

File: src/lib/util/logging.ts (hunk 1 above)
Comment (ruowan) [reply src/lib/util/logging.ts:22]: toString()

File: src/lib/util/resolveSwagger.ts (hunk 2 above)
Comment (jianyexi) [reply src/lib/util/resolveSwagger.ts:64]: fixed

File: src/lib/util/logging.ts (hunk 1 above)
Comment (jianyexi) [reply src/lib/util/logging.ts:22]: Good point

File: src/lib/util/resolveSwagger.ts (hunk 3 above)
Comment (jianyexi) [reply src/lib/util/resolveSwagger.ts:129]: fixed

File: src/lib/util/resolveSwagger.ts (hunk 7 above)
Comment (jianyexi) [reply src/lib/util/resolveSwagger.ts:200]: added

File: src/lib/util/resolveSwagger.ts (hunk 4 above)
Comment (jianyexi) [reply src/lib/util/resolveSwagger.ts:253]: added

File: src/lib/util/resolveSwagger.ts (hunk 5 above)
Comment (jianyexi) [reply src/lib/util/resolveSwagger.ts:267]: Good point.

**Review by ruowan — APPROVED** [review]

**Pull request merged**